libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -268,6 +268,61 @@ void ContractCompiler::appendDelegatecallCheck()
 	// "We have not been called via DELEGATECALL".
 }
 
+void ContractCompiler::appendInternalSelector(
+	map<FixedHash<4>, eth::AssemblyItem const> const& _entryPoints,
+	vector<FixedHash<4>> const& _ids,
+	eth::AssemblyItem const& _notFoundTag,
+	size_t _runs
+)
+{
+	// Code for selecting from n functions without split:
+	//   n times: dup1, push4 <id_i>, eq, push2/3 <tag_i>, jumpi
+	//   push2/3 <notfound> jump
+	// (called SELECT[n])
+	// Code for selecting from n functions with split:
+	//   dup1, push4 <pivot>, gt, push2/3<tag_less>, jumpi
+	//     SELECT[n/2]
+	//   tag_less:
+	//     SELECT[n/2]
+	//
+	// This means each split adds 16-18 bytes of additional code (note the additional jump out!)
+	// The average execution cost if we do not split at all are:
+	//   (3 + 3 + 3 + 3 + 10) * n/2 = 24 * n/2 = 12 * n
+	// If we split once:
+	//    (3 + 3 + 3 + 3 + 10) + 24 * n/4 = 24 * (n/4 + 1) = 6 * n + 24;
+	//
+	// We should not split if
+	//     _runs * 12 * n <= _runs * (6 * n + 24) + 17 * createDataGas
+	// <=> _runs * 6 * (n - 4) <= 17 * createDataGas
+	//
+	// Which also means that the execution itself is not profitable
+	// unless we have at least 5 functions.
+
+	if (_ids.size() <= 4 || _runs * 6 * (_ids.size() - 4) <= 17 * eth::GasCosts::createDataGas)
```

Comment:
Thinking about this: should we limit the value of ``_runs`` to something reasonably high? If someone thinks "my contract will be used and called a lot for all eternity", one might attempt to choose an "infinite-like" value like ``2^64-1`` which will overflow the left-hand side here.

Reply:
We could e.g. consider everything starting from a billion as "infinite" and always use binary search then. Or a million thinking about 32-bit systems.

Reply:
Or actually check for an overflow here, that'd maybe be even cleaner, since we also don't know ``_ids.size()``.

Reply:
Could we just check for some sane values here? I.e. if `_runs = 0` or `_ids.size() <= 4`, then don't split. If `_runs > datagas * 17 / 6`, then split. Otherwise, do the calculation.

Reply:
Updated.